numpyro/distributions/continuous.py

```diff
@@ -1188,6 +1188,70 @@ def infer_shapes(
         return batch_shape, event_shape
 
 
+class CAR(Distribution):
+    arg_constraints = {
+        "loc": constraints.real_vector,
+        "alpha": constraints.unit_interval,
+        "tau": constraints.positive,
+        "W": constraints.positive_definite,
```

Comment:
I think you can skip the `W` constraint for now (so we can allow the sparse case and this distribution is fully reparametrized). We can revisit this in the future if it is worth to introduce a new Constraint object for `W`.

Reply:
I'm confused. Are the users going to input a sparse W, or are we going to convert W to a sparse matrix within the log_p block? I thought the latter

Reply:
I thought we allow users to specify sparse W to save memory and time complexity.

Are you going to add a flag or a check for sparsity to convert a dense W to a sparse W? If we follow that path, I'm not sure what's a proper way to do such conversion. Seems like we need some check that W is a static numpy array, then use some scipy dense_to_sparse operator on the top of that (those operators should not operate on jax arrays because we need a static size of sparse form of W). Though if I'm a user of the sparse version, I'm not sure why I need to provide the dense version. (I think both stan and pymc allow users to provide the sparse version)

Reply:
Hmm. I was thinking to put something like 

`self.sparse = True if (1.0 - np.count_nonzero(np.array(self.W))/self.W.size) > 0.5`. Or maybe there's another numpy/scipy function which can determine this. Basically something that determines if W is sparse. But if we want to allow the user to decide whether to use sparse computations regardless of whether W is actually sparse, we could have an extra argument sparse, and a flag if they input W as anything other than a sparse matrix when sparse=True.

Would this work with our implementation of sample though?

Reply:
I see. Please go ahead if you want to use the dense matrix as the input and use some heuristic rule for the conversion. It is less flexible for users but it should work, just make sure to check that W is a numpy array before applying those numpy operators. In the future, we can add CARsparse or something like that if users want to use a sparse version. I dont know how to derive a sparse formula `sample` method - we might just use the dense version there.

I think stan also requires users to provide sparse W for the car prior. It does some sorts of data transformation from dense to sparse in advance, which is a reasonable approach for a tutorial (in practice, assume that a sparse W is stored in some csv file, that csv should only store the sparse version for memory and space efficient).

Reply:
> something that determines if W is sparse

I would like to control this by myself, rather than relying on some heuristic rule. It is hard to decide which threshold to use (given different performance on the shapes, the backends cpu tpu gpu, the memory,...) Even if I have a dense array from somewhere, I would like to call dense_to_sparse outside of mcmc code. The signature is the same with the cost of one additional line of code for that conversion (it would cost nothing if I already have a sparse array). As above, I can just simply create a new CARsparse version if I want such behavior. So please just go ahead with the heuristic approach.

Reply:
We can create a `CARsparse` separately if you think it's easier for the user? If we can work out how to sample from MVN with a sparse precision matrix, it won't be much more work. Up to you

Reply:
Actually, we can have a compromised version for all usage cases:
def init(W, is_sparse=False):
if is_sparse:
    if not scipy...issparse(W):
        assert isinstance(W, np.ndarray)
        W = scipy.sparse....from_dense(W)

Reply:
Got you. So let the user decide whether to use sparse computations, but also let the user decide whether to input sparse W or not. Perfect.

In `sample`, shall we just stick to dense computations?

Reply:
> shall we just stick to dense computations

I think so. There should be a way to utilize the sparse form but I can't derive it now. Unless you want to perform variational inference on the top of this prior, otherwise, `sample` is not required for MCMC.

Reply:
There are a [bunch](https://docs.scipy.org/doc/scipy-1.8.0/html-scipyorg/reference/sparse.html) of sparse matrix formats. What's the best way of coping if a user inputs the wrong type?

Reply:
Also, what actually is scale_tril? I'm assuming it's not just the lower diagonal of a covariance matrix

Reply:
I think you can use `W = W.tocsr()` to make sure it has the desired sparse format to perform multiplication for `W_scaled`. `scale_tril` is the cholesky of covariance matrix.

Reply:
There's a [sparse cholesky from sci-kit](https://scikit-sparse.readthedocs.io/en/latest/cholmod.html), but is it usable, and is it worth using for `scale_tril`?

Reply:
I'm not sure. I guess we are trying to compute cholesky of inverse of sparse matrix here.

Reply:
Scipy has a sparse inverse. At this stage it isn't worth it. When the spatial modellers start using the function in practice, we'll see if this is a pinch point and can evaluate whether it is worth implementing

Reply:
agreed!